c2rust-analyze/src/rewrite/expr/hir_op.rs

```diff
@@ -0,0 +1,392 @@
+                        let arr = self.get_subexpr(ex, 0);
+                        let idx = Rewrite::CastUsize(Box::new(self.get_subexpr(ex, 1)));
+                        let elem = Rewrite::SliceTail(Box::new(arr), Box::new(idx));
+                        hir_rw = Rewrite::Ref(Box::new(elem), mutbl_from_bool(mutbl));
+                    }
+
+                    mir_op::RewriteKind::SliceFirst { mutbl } => {
+                        // `p` -> `&p[0]`
+                        let arr = hir_rw;
+                        let idx = Rewrite::LitZero;
+                        let elem = Rewrite::Index(Box::new(arr), Box::new(idx));
+                        hir_rw = Rewrite::Ref(Box::new(elem), mutbl_from_bool(mutbl));
+                    }
+
+                    mir_op::RewriteKind::MutToImm => {
+                        // `p` -> `&*p`
+                        let place = Rewrite::Deref(Box::new(hir_rw));
+                        hir_rw = Rewrite::Ref(Box::new(place), hir::Mutability::Not);
+                    }
+
+                    mir_op::RewriteKind::RemoveAsPtr => {
+                        assert!(matches!(hir_rw, Rewrite::Identity));
+                        //assert_eq!(num_args, 1);
+                        hir_rw = self.get_subexpr(ex, 0);
+                    }
+                }
+            }
+        }
+
+        // Emit rewrites on subexpressions first.
+        let applied_mir_rewrite = !matches!(hir_rw, Rewrite::Identity);
+        self.with_materialize_adjustments(applied_mir_rewrite, |this| {
+            intravisit::walk_expr(this, ex);
+        });
+
+        // Materialize adjustments if requested by an ancestor.
+        if self.materialize_adjustments {
+            let adjusts = self.typeck_results.expr_adjustments(ex);
+            assert!(
+                adjusts.is_empty() || !applied_mir_rewrite,
+                "combining adjustments ({:?}) with rewrite ({:?}) is NYI",
+                adjusts,
+                hir_rw,
+            );
+            for adj in adjusts {
+                match adj.kind {
+                    Adjust::NeverToAny => {
+                        // Should work fine with no explicit cast.
+                    }
+                    Adjust::Deref(_) => {
+                        hir_rw = Rewrite::Deref(Box::new(hir_rw));
+                    }
+                    Adjust::Borrow(AutoBorrow::Ref(_, mutbl)) => {
+                        hir_rw = Rewrite::Ref(Box::new(hir_rw), mutbl.into());
+                    }
+                    Adjust::Borrow(AutoBorrow::RawPtr(mutbl)) => {
+                        hir_rw = Rewrite::AddrOf(Box::new(hir_rw), mutbl.into());
+                    }
+                    Adjust::Pointer(PointerCast::Unsize) => {
+                        // TODO: figure out what kind of unsize this is and insert an explicit cast
+                        // (e.g. `&x` -> `&x as &[_]`).  In many cases this should work without a cast.
```

Comment:
Should this have a `todo!()`?

Reply:
i'm hesitant to add new places that can panic given our stated intention of making static analysis more robust (i.e. not crashing) for unsupported features

Reply:
Yeah, that's better actually.  Maybe just an `eprintln!` warning then?